### ISSUE TITLE
N

### ISSUE THREAD

**Issue by swapyface** [swapyface/test#2]
import pandas as pd
import numpy as np
from sklearn.mixture import GaussianMixture
from sklearn.preprocessing import StandardScaler, OneHotEncoder
import matplotlib.pyplot as plt

# Example dataset (replace with your actual dataset)
data = pd.DataFrame({
    'price': [500, 520, 550, 1000, 1020, 1500, 2000, 2500, 3000, 3200, 3500, 4000, 4500, 5000, 6000, 7000],
    'rating': ['Low', 'Low', 'Medium', 'Medium', 'High', 'WR', 'NR', 'Low', 'Medium', 'High', 'unknown', 'High', 'Low', 'Medium', 'High', 'Low']
})

# Directly flag records with 'WR', 'NR', or 'unknown' ratings as outliers (True)
data['manual_flag'] = data['rating'].isin(['WR', 'NR', 'unknown'])
data['anomaly_level'] = np.where(data['manual_flag'], 'High', np.nan)

# Handle missing values: fill missing prices with 0 and missing ratings with 'unknown'
data['price'].fillna(0, inplace=True)
data['rating'].fillna('unknown', inplace=True)

# Separate the records that are flagged manually
flagged_data = data[data['manual_flag']]
non_flagged_data = data[~data['manual_flag']]

# One-hot encode the categorical variable 'rating' (excluding the manually flagged)
encoder = OneHotEncoder(sparse=False)
rating_encoded = encoder.fit_transform(non_flagged_data[['rating']])
encoded_df = pd.DataFrame(rating_encoded, columns=encoder.get_feature_names_out(['rating']))

# Concatenate encoded categorical variables with the original dataset
non_flagged_data = pd.concat([non_flagged_data, encoded_df], axis=1)

# Scale the numerical data (important for GMM)
scaler = StandardScaler()
non_flagged_data['scaled_price'] = scaler.fit_transform(non_flagged_data[['price']])

# Apply GMM (Gaussian Mixture Model)
gmm = GaussianMixture(n_components=3, random_state=42)  # Adjust 'n_components' as needed
gmm.fit(non_flagged_data[['scaled_price'] + list(encoded_df.columns)])
non_flagged_data['gmm_anomaly_score'] = gmm.score_samples(non_flagged_data[['scaled_price'] + list(encoded_df.columns)])

# Set thresholds for categorizing anomaly scores
non_flagged_data['anomaly_level'] = pd.cut(non_flagged_data['gmm_anomaly_score'], 
                                           bins=[-np.inf, -0.7, -0.5, np.inf], 
                                           labels=['High', 'Medium', 'Low'])

# Combine the manually flagged anomalies with the GMM results
final_data = pd.concat([flagged_data, non_flagged_data])

# Save the final dataset to CSV
final_data.to_csv('gmm_anomalies_with_manual_flags.csv', index=False)

# Visualize the results using a scatter plot
plt.figure(figsize=(10, 6))
plt.scatter(final_data.index, final_data['price'], c=final_data['anomaly_level'].apply(lambda x: 'red' if x == 'High' else ('orange' if x == 'Medium' else 'blue')))
plt.title('Anomaly Detection with Manual Flags and GMM')
plt.xlabel('Index')
plt.ylabel('Price')
plt.show()

import pandas as pd
import numpy as np
from sklearn.mixture import GaussianMixture
from sklearn.preprocessing import StandardScaler, OneHotEncoder
import matplotlib.pyplot as plt

# Example dataset (replace with your actual dataset)
data = pd.DataFrame({
    'price': [500, 520, 550, 1000, 1020, 1500, 2000, 2500, 3000, 3200, 3500, 4000, 4500, 5000, 6000, 7000],
    'rating': ['Low', 'Low', 'Medium', 'Medium', 'High', 'WR', 'NR', 'Low', 'Medium', 'High', 'unknown', 'High', 'Low', 'Medium', 'High', 'Low']
})

# Directly flag records with 'WR', 'NR', or 'unknown' ratings as outliers (True)
data['manual_flag'] = data['rating'].isin(['WR', 'NR', 'unknown'])
data['Is_outlier'] = np.where(data['manual_flag'], True, False)  # Initialize the 'Is_outlier' column

# Handle missing values: fill missing prices with 0 and missing ratings with 'unknown'
data['price'].fillna(0, inplace=True)
data['rating'].fillna('unknown', inplace=True)

# Separate the records that are flagged manually
flagged_data = data[data['manual_flag']]
non_flagged_data = data[~data['manual_flag']]

# One-hot encode the categorical variable 'rating' (excluding the manually flagged)
encoder = OneHotEncoder(sparse=False)
rating_encoded = encoder.fit_transform(non_flagged_data[['rating']])
encoded_df = pd.DataFrame(rating_encoded, columns=encoder.get_feature_names_out(['rating']))

# Concatenate encoded categorical variables with the original dataset
non_flagged_data = pd.concat([non_flagged_data, encoded_df], axis=1)

# Scale the numerical data (important for GMM)
scaler = StandardScaler()
non_flagged_data['scaled_price'] = scaler.fit_transform(non_flagged_data[['price']])

# Apply GMM (Gaussian Mixture Model)
gmm = GaussianMixture(n_components=2, random_state=42)  # Adjust 'n_components' based on your dataset
gmm.fit(non_flagged_data[['scaled_price'] + list(encoded_df.columns)])
non_flagged_data['gmm_cluster'] = gmm.predict(non_flagged_data[['scaled_price'] + list(encoded_df.columns)])

# Map GMM clusters to inliers (0) and outliers (1)
# We assume that the cluster with fewer points is the outlier cluster
cluster_sizes = non_flagged_data['gmm_cluster'].value_counts()
outlier_cluster = cluster_sizes.idxmin()  # Outlier cluster is the smaller cluster

non_flagged_data['Is_outlier'] = non_flagged_data['gmm_cluster'] == outlier_cluster

# Combine the manually flagged anomalies with the GMM results
final_data = pd.concat([flagged_data, non_flagged_data])

# Save the final dataset to CSV
final_data.to_csv('gmm_inliers_outliers_with_manual_flags.csv', index=False)

# Visualize the results using a scatter plot
plt.figure(figsize=(10, 6))
plt.scatter(final_data.index, final_data['price'], c=final_data['Is_outlier'].apply(lambda x: 'red' if x else 'blue'))
plt.title('Inliers vs Outliers (GMM with Manual Flags)')
plt.xlabel('Index')
plt.ylabel('Price')
plt.show()
